text/0322-cdk-pipelines-updated-api.md

```diff
@@ -0,0 +1,324 @@
+    repo: 'REPO',
+    trigger: codepipeline_actions.GitHubTrigger.POLL,
+  }),
+
+  // How it will be built
+  synthAction: cdkp.SimpleSynthAction.standardNpmSynth({
+    sourceArtifact,
+    cloudAssemblyArtifact,
+    projectName: 'MyServicePipeline-synth',
+    additionalArtifacts: [
+      {
+        directory: 'test',
+        artifact: integTestArtifact,
+      },
+    ],
+  }),
+});
+
+const stage = new MyStage(this, 'PreProd', {
+  env: { account: '12345', region: 'us-east-1' },
+});
+stage.addActions(
+  new cdkp.ShellScriptAction({
+    commands: ['node ./integ-tests'],
+    additionalArtifacts: [integTestArtifact],
+  }),
+);
+```
+
+Becomes:
+
+```ts
+import * as rollout from '@aws-cdk/rollout';
+
+const pipeline = new rollout.Rollout(this, 'Pipeline', {
+  build: rollout.Build.shellScript({
+    input: rollout.CodePipelineSource.gitHub('OWNER/REPO'),
+    commands: ['npm ci', 'npm run build'],
+    additionalOutputs: {
+      tests: rollout.AdditionalOutput.fromDirectory('test'),
+    }
+  }),
+  backend: new rollout.AwsCodePipelineBackend(),
+});
+
+const stage = new MyStage(this, 'PreProd', {
+  env: { account: '12345', region: 'us-east-1' },
+});
+pipeline.addApplicationStage(stage, {
+  approvals: [
+    rollout.Approval.shellScript({
+      input: pipeline.build.additionalOutput('tests'),
+      commands: ['node ./integ-tests'],
+    }),
+  ],
+});
+```
+
+### How we customize CodeBuild projects, and other things that are AWS-specific?
+
+AWS-specific customizations are passed as parameters to the Backend class:
+
+```
+const pipeline = new rollout.Rollout(this, 'Pipeline', {
+  build: rollout.Build.shellScript({
+    input: rollout.CodePipelineSource.gitHub('OWNER/REPO'),
+    commands: ['npm ci', 'npm run build', 'npx cdk synth'],
+    environment: {
+      NPM_CONFIG_UNSAFE_PERM: 'true',
+    },
+  }),
+  backend: new rollout.AwsCodePipelineBackend({
+    pipeline,
+    pipelineName: 'MyPipeline',
+    vpc,
+    subnetSelection: { subnetType: ec2.SubnetType.PRIVATE },
+    crossAccountKeys: true,
+    buildEnvironment: {
+      image: codebuild.CodeBuildImage.AWS_STANDARD_6,
+      privilegedMode: true,
+    },
+    buildCaching: true,
+    cdkCliVersion: '1.2.3',
+    selfMutating: false,
+    pipelineUsesDockerAssets: true,
+    dockerAssetBuildPrefetch: true,
+    dockerCredentials: {
+      '': rollout.DockerCredentials.fromSecretsManager('my-dockerhub-login'),
+      '111111.dkr.ecr.us-east-2.amazonaws.com': rollout.DockerCredentials.standardEcrCredentials(),
+    },
+    buildTestReports: {
+      SurefireReports: {
+        baseDirectory: 'target/surefire-reports',
+        files: ['**/*'],
+      }
+    },
+  }),
+});
+```
+
+## Implementation FAQ
+
+### How will this work?
+
+The library will be organized into 3 layers:
+
+```
+┌──────────────────┬────────────────┬───────────────────────┐
+│                  │                │    Generic actions    │
+│     Backend,     │    Rollout     │                       │
```

Comment:
Do we really need to separate rollout from the workflow core now?

Reply:
Depends on how you want to look at it. In my head, those are data structures that the user doesn't necessarily directly interact with (the front-level API will be mostly in terms of CDK apps, which get translated into middle-level Actions).

Reply:
I think I understand. So basically the middle layer is the model from which the backends render their concrete implementation. Did I get that right?